docs: better `global ignores` instruction (#19297)
eslint/eslint: docs/src/use/configure/ignore.md

This page explains how to ignore files using the flat config format. For the deprecated eslintrc format, [see the deprecated documentation](ignore-deprecated).
:::

::: tip

> This page explains how to use `ignores` property of an ESLint configuration object. The `ignores` property can behave as global or non-global, based on how you use it. Please read how [ignores can be used as global or non-global](configuration-files) before going through this page.

Reviewer: ```suggestion
This page explains how to use `ignores` property of an ESLint configuration object to globally ignore files and directories. The `ignores` property can behave as either global or non-global depending on how you use it. For more information on non-global ignores, see [Specifying files and ignores](configuration-files#specifying-files-and-ignores).
```

Author: Added in https://github.com/eslint/eslint/pull/19297/commits/45ae8059c9807d02a53f471f1bc9a47b38c63997 with an additional line.

> ::: tip
> This page explains how to use `ignores` property of an ESLint configuration object. The `ignores` property can behave as global or non-global, based on how you use it. Please read how [ignores can be used as global or non-global](configuration-files#globally-ignoring-files-with-ignores) before going through this page.
> :::
> 

Reviewer 2: ```suggestion
::: tip
The `ignores` property in ESLint configuration object can behave as global or non-global, based on how you use it. Please read how [ignores can be used as global or non-global](configuration-files#globally-ignoring-files-with-ignores) before going through this page.
:::

```

first sentence seems repetitive, how about if we write something like this?

];
```

> ## Global Ignores

Reviewer: This section seems out of place with the rest of the content on this page. The entire page is about ignoring files and directories using global ignores, so it doesn't make sense to have a separate section.

I also feel like some of this content is redundant to what's on the configuration files page.

I'd much rather see some of this content distributed throughout this page in the existing sections rather than putting in an entirely new block that duplicates some of the content.

Author: Yeah! There is a lot of ambiguities between these two pages.

Personally, if I start reading [Ignore](https://deploy-preview-19297--docs-eslint.netlify.app/use/configure/ignore) page I don't think this page is related to "global" only configuration of ignores. But to "ignores" in general.

For me the simpler approach for the reader is to have:
 - [Configuration Files](https://deploy-preview-19297--docs-eslint.netlify.app/use/configure/configuration-files#configuration-objects) page with a brief description of every properties of configuration object, with links to other pages if that property need big explanation (like ignores) <img width="400" alt="Schermata 2025-01-02 alle 20 14 29" src="https://github.com/user-attachments/assets/4e75651b-278e-4c04-a7f7-76d7608d0d53" />
- move all these subsections related to ignores from **Configuration Files** page to **Ignores** page <img width="157" alt="Schermata 2025-01-02 alle 20 18 43" src="https://github.com/user-attachments/assets/46f6c1b9-4be5-4c02-9aa8-b759390cc60f" />
- remove redundancy and refactor sections of **Ignores** page due to big amount of information

This is not an easy task, this is why i marked this as **Out of scpoe of this PR**.
What do you think ? @nzakas @Tanujkanti4441

Reviewer: I agree that we don't want to completely reorganize the documentation for this PR. There are a lot of links to different sections that would need to be verified to avoid breaks, so I think for this PR we want to focus on building out content within the current structure.

Author: Great! 
Do you have an idea of how to proceed?

I haven't an idea of how to refactor this piece of text (the one attached to this comment )

Reviewer: I don't have the time to look into that right now. To move forward, I'd suggest just reverting the changes to `ignore.md` so we can merge the rest of this PR. We can revisit additional changes later.

Author: I agree!

Changes made:

- in https://github.com/eslint/eslint/pull/19297/commits/00b058504034df2d48089ebcc664f8fe87b655ba : 
reverted all changes to `ignore.md`, now the ignore page is at the same state "before this PR"
- in https://github.com/eslint/eslint/pull/19297/commits/00b058504034df2d48089ebcc664f8fe87b655ba : 
[Live Preview](https://deploy-preview-19297--docs-eslint.netlify.app/use/configure/ignore)
To inform the user who arrives on `ignore.md` page without having gone through `configuration-files.md` (where we explain the difference between global and non-global ignores) I added a `tip/alert` at the top of `ignore.md` page, hoping the user will read it before going through the `ignore.md` page. <img width="540" alt="Schermata 2025-01-07 alle 20 30 22" src="https://github.com/user-attachments/assets/a235421c-4b9b-4bf1-80d3-8b2c1635f1e1" />

**Please check english syntax/grammar of this tip/alert**, I'm not a native english speaker

:::

In eslint, there is concept of `global ignores`, to better explain it let's see first how config object works in eslint.

Eslint configuration file is intended to be used like this: export a single array, that contains multiple [configuration objects](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects), each one being an item of the array.  

It's important to note that each configuration object, it's independent from other ones. Because of that, the `ignores` prop defined in one configuration object is not applied to other configuration objects of the array.

```js
// eslint.config.js

export default [
    { name: "config-1", files: "src/*", ignores: [".dist/*"], ... }, 
    { name: "config-2", files: "scripts/*", ignores: [".others/*"], ... }, 
    // ...
];
```

In the code above, the `ignores` directive of `config-1` is not applied on `config-2`, and vice-versa.  If this is what you want, use this pattern.

But usually, what you want is to globally ignores some directory/files, and apply this directive to all configuration objects of the array:  `global ignores` is exactly made for this scenario.

> The usage is the following:
> - define a configuration object with a single prop, `ignores` .
> - **don't add any other props** to this configuration object, only `ignores` prop
> - place this configuration objects as **first item** of the configuration object arrays
> 

Reviewer 2: ```suggestion
The usage is the following:
* define a configuration object with a single prop, `ignores` .
* **don't add any other props** to this configuration object, only `ignores` prop
* place this configuration objects as **first item** of the configuration object arrays

```

To fix the CI errors

];
```

## Global Ignores

::: tip
Global Ignores is what you need if you want to define files to ignore in a single place and avoid to add the same `ignores` prop to every configuration objects.
:::

> In eslint, there is concept of `global ignores`, to better explain it let's see first how config object works in eslint.
> 
> Eslint configuration file is intended to be used like this: export a single array, that contains multiple [configuration objects](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects), each one being an item of the array.  
> 
> It's important to note that each configuration object, it's independent from other ones. Because of that, the `ignores` prop defined in one configuration object is not applied to other configuration objects of the array.

Reviewer 2: ```suggestion
In ESLint, there is a concept of global `ignores`, to better explain it let's see first how configuration object works in ESLint.

ESLint configuration file is intended to be used like this: export a single array, that contains multiple [configuration objects](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects), each one being an item of the array.  

It's important to note that each configuration object is independent from other ones. Because of that, the `ignores` property defined in one configuration object is not applied to other configuration objects of the array.
```

```js
// eslint.config.js

> export default [
>     { name: "config-1", files: "src/*", ignores: [".dist/*"], ... }, 
>     { name: "config-2", files: "scripts/*", ignores: [".others/*"], ... }, 
>     // ...
> ];

Reviewer 2: ```suggestion
export default [
    { name: "config-1", files: "src/**/*.js", ignores: ["**/*.config.js"], ... }, 
    { name: "config-2", files: "scripts/**/*.js", ignores: ["**/dir/**"], ... }, 
    // ...
];
```

Here we can show that `config-1` is ignoring a file and `config-2` is ignoring files inside a directory.

```

> In the code above, the `ignores` directive of `config-1` is not applied on `config-2`, and vice-versa.  If this is what you want, use this pattern.

Author: Is true this part ? 

I'm not sure that what i wrote it's true, this is my personal assumption based on my usage of eslint, but I'm not an expert in eslint.
I want to be sure to not create more confusion...

Author: I'm referring specifically to 

> It's important to note that each configuration object, it's independent from other ones. Because of that, the `ignores` prop defined in one configuration object is not applied to other configuration objects of the array.

Configuration objects are independent?

Reviewer 2: Yes they are.

Reviewer 2: yes they are.

> But usually, what you want is to globally ignores some directory/files, and apply this directive to all configuration objects of the array:  `global ignores` is exactly made for this scenario.

Reviewer 2: ```suggestion
But if you want to apply a particular `ignores` pattern to all configuration objects of the configuration array,  global `ignores` is exactly made for this scenario.
```

The usage is the following:

> * define a configuration object with a single prop, `ignores` .
> * **don't add any other props** to this configuration object, only `ignores` prop
> * place this configuration objects as **first item** of the configuration object arrays

Reviewer 2: ```suggestion
* define a configuration object with a single property, `ignores` .
* **don't add any other property** to this configuration object other than `ignores` property.
* place this configuration object as **first item** of the configuration array.
```

> This way every configuration object that follow this one inherit the `global ignores`.

Reviewer 2: ```suggestion
This way every configuration object that follow this one inherit the global `ignores`.
```

> 
> ```ts

Reviewer 2: ```suggestion

```js
```

Let's keep it a `js` example

> // eslint.config.ts
> 
> import globals from "globals";
> import pluginJs from "@eslint/js";
> import tseslint from "typescript-eslint";
> 
> /** @type {import('eslint').Linter.Config[]} */
> export default [
>   // DO NOT PUT OTHER PROPS IN THIS OBJECT, SO IT
>   // BEHAVES AS GLOBAL IGNORES
>   { ignores: ["**/node_modules/**", "dist/"] }, 
> 
>   // THE IGNORES IS APPLIED ALSO TO THESE FOLLOWING CONFIGS
>   { files: ["**/*.{js,mjs,cjs,ts}"] },
>   { languageOptions: { globals: globals.browser } },
>   pluginJs.configs.recommended,
>   ...tseslint.configs.recommended,
> ];

Reviewer 2: ```suggestion
// eslint.config.js

import globals from "globals";
import pluginJs from "@eslint/js";
import tseslint from "typescript-eslint";

export default [
  // DO NOT PUT OTHER PROPS IN THIS OBJECT, SO IT
  // BEHAVES AS GLOBAL IGNORES
  { ignores: ["**/node_modules/**", "dist/"] }, 

  // THE IGNORES IS APPLIED ALSO TO THESE FOLLOWING CONFIGS
  { files: ["**/*.{js,mjs,cjs,ts}"] },
  { languageOptions: { globals: globals.browser } },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
];
```
